cpp/src/lists/lists_column_factories.cu

```diff
@@ -104,15 +105,25 @@ std::unique_ptr<column> make_lists_column(size_type num_rows,
   CUDF_EXPECTS(offsets_column->null_count() == 0, "Offsets column should not contain nulls");
   CUDF_EXPECTS(child_column != nullptr, "Must pass a valid child column");
 
+  // Save type_id of the child column for later use.
+  auto const child_type_id = child_column->type().id();
+
   std::vector<std::unique_ptr<column>> children;
   children.emplace_back(std::move(offsets_column));
   children.emplace_back(std::move(child_column));
-  return std::make_unique<column>(cudf::data_type{type_id::LIST},
-                                  num_rows,
-                                  rmm::device_buffer{},
-                                  std::move(null_mask),
-                                  null_count,
-                                  std::move(children));
+
+  auto output = std::make_unique<column>(cudf::data_type{type_id::LIST},
+                                         num_rows,
+                                         rmm::device_buffer{},
+                                         std::move(null_mask),
+                                         null_count,
+                                         std::move(children));
+
+  // We need to enforce all null lists to be empty.
+  // Checking `null_count==0` is not enough as it can be `UNKNOWN_NULL_COUNT` for nullable column.
+  return null_count == 0 || !output->nullable() || child_type_id == type_id::EMPTY
+           ? std::move(output)
+           : detail::purge_nonempty_nulls(output->view(), stream, mr);
```

Comment:
Are there concerns about the performance implications here where we have gone from a move to a copy in the case of calling `purge_nonempty_nulls`?

Reply:
`purge_nonempty_nulls` is a very trivial gathering/memcpy thus it should have very small performance impact.